proposals/kapp/001-crd-upgrade-safety/README.md

```diff
@@ -0,0 +1,95 @@
+---
+title: "Adding CRD Upgrade Safety to kapp"
+authors: [ "Bryce Palmer <bpalmer@redhat.com>" ]
+status: "in review"
+approvers: [ "Carvel Maintainers" ]
+---
+
+# Adding CRD Upgrade Safety to kapp
+
+## Problem Statement
+CustomResourceDefinitions (CRDs) are an integral part of extending the Kubernetes API and are often applied to a cluster alongside a controller that manages them.
+
+Since CRDs are global and only one instance can exist on the cluster at a time, it is important to ensure that updates to a CRD are performed safely.
+This means preventing data loss and breaking of clients/workloads that may be using the API.
+Failure to safely update a CRD may result in:
+- Existing CustomResources (CRs) being invalidated, causing extensions to fail reconciliation of the existing CRs
+- Breaking clients expecting the old schema to be present. This may include other extensions, workloads, and tooling that depend on a specific version of the CRD.
+
+When an unsafe update occurs, it requires manual intervention to update all of the existing CRs to the new version of the CRD.
+
+## Terminology / Concepts
+- CustomResourceDefinition (CRD) - https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/
+- CustomResource (CR) - https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/
+- Fail Closed - In the scope of this document, "failing closed" is used to represent that if we can't explicitly validate something then we should fail.
+- Fail Open - In the scope of this document, "failing open" is used to represent that if we can't explicitly validate something then we should succeed.
+- Preflight Check - A preflight check is a validation that is run prior to performing some action. Typically, if the validation fails the action will not be performed.
+
+## Proposal
+In order to address the problem identified, it is proposed that a new, optional, preflight check is added to `kapp` to enable verifying that an `update` operation on a CRD is safe before attempting to apply any changes to the cluster.
```

Comment:
Should we call `patch` out here as well?

Reply:
IIRC kapp doesn't currently perform any `patch` operations, meaning that for now we don't need to be concerned with it.

Reply:
This comment has been sitting for a bit. Going to mark as resolved. Please feel free to unresolve if you feel this has been insufficiently addressed.